File: util/testcase.py
#!/usr/bin/env python3

from dotenv import load_dotenv, find_dotenv
load_dotenv(find_dotenv())

import sys
import os
import json
from munch import Munch
from types import SimpleNamespace
from glob import glob

from compact_json import Formatter, EolStyle
formatter = Formatter()
formatter.indent_spaces = 2
formatter.max_inline_complexity = 10

from subprocess import call

import argparse
#import shlex
#import sys
from github import Github
#import os
import re
#from pathlib import Path
import shutil
import subprocess
#import string

from pygit2 import Repository
from gherkin.parser import Parser
from pathvalidate import sanitize_filename

root = os.path.join(os.path.dirname(__file__), '..')
try:
  branch = Repository(root).head.shorthand
  issue = int(branch.replace('gh-', ''))
except ValueError:
  issue = None

# get arguments
parser = argparse.ArgumentParser()
parser.add_argument('--translator', '-t', dest='translator', default='biblatex')
parser.add_argument('--data',       '-d', dest='data', required=True)
parser.add_argument('--feature',    '-f', dest='feature')
parser.add_argument('--issue',      '-i', dest='issue', default=issue and str(issue))
parser.add_argument('--export',     '-e', dest='mode', action='store_const', const='export')
parser.add_argument('--import'          , dest='mode', action='store_const', const='import')
parser.add_argument('--citekey'         , dest='mode', action='store_const', const='citekey')
parser.add_argument('--prefs'           , action='store_true')
parser.add_argument('--attachments'     , action='store_true')
parser.add_argument('--attach'     , action='store_true')
args, unknownargs = parser.parse_known_args()
sys.argv = sys.argv[:1] + unknownargs

if not args.mode:
  args.mode = 'export'

if not args.feature:
  args.feature = os.path.join(root, 'test', 'features', f'{args.mode}.feature')
assert os.path.exists(args.feature),  f'{args.feature} does not exist'

args.translator = {
  'biblatex': 'BibLaTeX',
  'bibtex': 'BibTeX',
  'csl': 'CSL-JSON',
  'csl-json': 'CSL-JSON',
  'yml': 'CSL-YAML',
  'csl-yaml': 'CSL-YAML',
}[args.translator.lower()]

assert args.issue, 'no issue'
args.data = os.path.join(root, 'logs', f'{args.data}/items.json')
assert os.path.exists(args.data),  f'{args.data} does not exist'

# get title

g = Github(os.environ['GITHUB_TOKEN'])
repo = g.get_repo('retorquere/zotero-better-bibtex')
issue = repo.get_issue(int(args.issue))
args.title = re.sub(r'^\[[^\]]+\]\s*:', '', issue.title).strip()
args.title = re.sub(r'^(Bug|Feature|Feature Request)\s*:', '', args.title, flags=re.IGNORECASE).strip()
args.title = re.sub(r'[^-A-Za-z0-9\s]', '', args.title).strip()
args.title = sanitize_filename(f'{args.title} #{issue.number}'.strip()).replace('`', '').replace('?', '')

if args.attach:
  candidates = { os.path.splitext(att)[1]: att for att in glob('attachments/*.*', root_dir = 'test/fixtures/export') }
  with open(args.data) as f:
    data = json.load(f)
    for item in data['items']:
      if 'attachments' in item:
        for att in item['attachments']:
          if 'path' in att:
            att['path'] = candidates[os.path.splitext(att['path'])[1]]
  with open(args.data, 'w') as f:
    print(formatter.serialize(data), file=f)

# clean lib before putting it in place
cleanlib = ["./util/clean-lib.ts", args.data]
if args.prefs: cleanlib.append('--prefs')
if args.attachments: cleanlib.append('--attachments')
assert call(cleanlib, cwd=root) == 0, 'clean failed'
with open(args.data) as f:
  data = json.load(f)
  args.n = len(data['items'])
  for item in data['items']:
    if 'attachments' in item:
      for att in item['attachments']:
        if 'path' in att:
          if not os.path.exists(os.path.join('test/fixtures/export', att['path'])):
            print('*** WARNING ***:', args.data, 'has non-existing attachment', att['path'])

# insert example
parser = Parser()
doc = Munch.fromDict(parser.parse(args.feature))

outlines = [
  child.scenario
  for child in doc.feature.children
  if child.get('scenario') and child.scenario.keyword == 'Scenario Outline' and (args.translator in child.scenario.name or args.mode == 'import')
]
assert len(outlines) == 1, f'{len(outlines)} outlines found containing {args.translator}'

with open(args.feature) as f:
  contents = f.readlines()
  example = [f'| {args.title} | {args.n} |\n']

examples = [example for example in outlines[0].examples[0].tableBody if example.cells[0].value == args.title]
match len(examples):
  case 0:
    # insert as first
    line = outlines[0].examples[0].tableHeader.location.line
    contents = contents[:line] + example + contents[line:]

  case 1:
    if examples[0].cells[1].value == str(args.n):
      contents = None
    else:
      # replace
      line = examples[0].location.line
      contents = contents[:line - 1] + example + contents[line:]
  case _:
    assert False, f'{len(examples)} examples found with title {json.dumps(args.title)}'

if contents:
  with open(args.feature, 'w') as f:
    f.write(''.join(contents))

# copy/create test fixtures
fixture = os.path.join(root, f'test/fixtures/{args.mode}', args.title)
source = fixture + '.json'
shutil.copyfile(args.data, source)
subprocess.check_output(['git', 'add', source])

if args.mode == 'import':
  ext = 'bib'
else:
  ext = args.translator.lower().replace('-', '.').replace('yaml', 'yml')

target = fixture + '.' + ext
with open(target, 'w') as f:
  if args.translator == 'CSL-JSON':
    f.write('{}')
subprocess.check_output(['git', 'add', target])

# reformat
sys.argv.append(args.feature)
from reformat_gherkin.cli import main
main()

File: util/testcase.mjs
#!/usr/bin/env node

import anyAscii from 'any-ascii'
import { execSync } from 'child_process'
import fs from 'fs'
import { globSync as glob } from 'glob'
import path from 'path'
import sanitize from 'sanitize-filename'
import { fileURLToPath } from 'url'

import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'

import { TableCell, TableRow } from 'gherkin-ast'
import { format } from 'gherkin-formatter'
import { read } from 'gherkin-io'

import { Octokit } from '@octokit/rest'

const root = path.join(path.dirname(fileURLToPath(import.meta.url)), '..')
import dotenv from 'dotenv'
dotenv.config({ quiet: true, path: path.join(root, '.env') })

// Get current Git branch and issue number
let issue = null
try {
  const branch = execSync('git rev-parse --abbrev-ref HEAD', { cwd: root, encoding: 'utf8' }).trim()
  const m = branch.match(/^gh-(?<issue>\d+)$/)
  if (m) issue = parseInt(m.groups.issue, 10)
}
catch (error) {
  issue = null
}

// Argument parsing
const argv = yargs(hideBin(process.argv))
  .option('translator', {
    alias: 't',
    type: 'string',
    default: 'biblatex',
    description: 'Translator type (biblatex, bibtex, csl, csl-json, yml, csl-yaml)',
  })
  .option('data', {
    alias: 'd',
    type: 'string',
    demandOption: true,
    description: 'Path to the data file',
  })
  .option('feature', {
    alias: 'f',
    type: 'string',
    description: 'Path to the feature file',
  })
  .option('issue', {
    alias: 'i',
    type: 'number',
    default: issue,
    description: 'GitHub issue number',
  })
  .option('title', {
    type: 'string',
    description: 'Scenario title',
  })
  .option('number', {
    alias: 'n',
    type: 'number',
    default: 1,
    description: 'Number of items',
  })
  .option('mode', {
    type: 'string',
    choices: ['import', 'export', 'citekey'],
    conflicts: ['import', 'export', 'citekey'],
    description: 'Set the script mode to import, export, or citekey',
  })
  .option('export', {
    alias: 'e',
    type: 'boolean',
    conflicts: ['mode', 'import', 'citekey'],
    description: 'Set mode to "export"',
  })
  .option('import', {
    type: 'boolean',
    conflicts: ['mode', 'export', 'citekey'],
    description: 'Set mode to "import"',
  })
  .option('citekey', {
    type: 'boolean',
    conflicts: ['mode', 'export', 'import'],
    description: 'Set mode to "citekey"',
  })
  .option('prefs', {
    type: 'boolean',
    description: 'Run with --prefs',
  })
  .option('attachments', {
    type: 'boolean',
    description: 'Run with --attachments',
  })
  .option('attach', {
    type: 'boolean',
    description: 'Attach files',
  })
  .parse()

if (argv.export) {
  argv.mode = 'export'
}
else if (argv.import) {
  argv.mode = 'import'
}
else if (argv.citekey) {
  argv.mode = 'citekey'
}
else if (!argv.mode) {
  argv.mode = 'export'
}
for (const mode of ['import', 'export', 'citekey']) {
  argv[mode] = argv.mode === mode
}
if (!argv.feature) argv.feature = path.join(root, 'test', 'features', `${argv.mode}.feature`)

const Translator = new class {
  constructor() {
    switch (argv.translator.toLowerCase()) {
      case 'biblatex':
        this.name = 'BibLaTeX'
        this.ext = '.biblatex'
        break
      case 'bibtex':
        this.name = 'BibTeX'
        this.ext = '.bibtex'
        break
      case 'csl':
      case 'csl-json':
        this.name = 'CSL-JSON'
        this.ext = '.csl.json'
        break
      case 'yml':
      case 'csl-yaml':
        this.name = 'CSL-YAML'
        this.ext = '.csl.yml'
        break
    }
  }
}()

const main = async () => {
  // Validate arguments
  if (!argv.issue) {
    console.error('Error: no issue number provided.')
    process.exit(1)
  }

  if (!fs.existsSync(argv.feature)) {
    console.error(`Error: Feature file does not exist at ${argv.feature}`)
    process.exit(1)
  }

  argv.data = path.join(root, 'logs', `${argv.data}/items.json`)
  if (!fs.existsSync(argv.data)) {
    console.error(`Error: Data file does not exist at ${argv.data}`)
    process.exit(1)
  }

  if (!Translator.name) {
    console.error(`Error: Invalid translator specified: ${argv.translator}`)
    process.exit(1)
  }

  if (!argv.title) {
    // Get issue title from GitHub
    if (!process.env.GITHUB_TOKEN) {
      console.error('Error: GITHUB_TOKEN environment variable is not set.')
      process.exit(1)
    }
    const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN })
    try {
      const { data: issue } = await octokit.rest.issues.get({
        owner: 'retorquere',
        repo: 'zotero-better-bibtex',
        issue_number: argv.issue,
      })
      argv.title = issue.title
        .replace(/^\[[^\]]+\]\s*:/, '').trim()
        .replace(/^(Bug|Feature|Feature Request)\s*:/i, '')
        .trim()
    }
    catch (error) {
      console.error(`Error fetching issue title for #${argv.issue}:`, error.message)
      process.exit(1)
    }
  }

  // Sanitize title
  argv.title = anyAscii(sanitize(`${argv.title} #${argv.issue}`).replace(/[`'"?]/g, ''))

  // Handle attachment logic
  if (argv.attach) {
    try {
      const attachmentDir = path.join(root, 'test/fixtures/export')
      const candidates = glob('attachments/*.*', { cwd: attachmentDir }).reduce((acc, file) => {
        const ext = path.extname(file)
        acc[ext] = file
        return acc
      }, {})

      const data = JSON.parse(fs.readFileSync(argv.data, 'utf8'))
      data.items.forEach(item => {
        if (item.attachments) {
          item.attachments.forEach(att => {
            if (att.path) {
              const ext = path.extname(att.path)
              if (candidates[ext]) {
                att.path = candidates[ext]
              }
            }
          })
        }
      })
      fs.writeFileSync(argv.data, JSON.stringify(data, null, 2), 'utf8')
    }
    catch (error) {
      console.error('Error handling attachments:', error.message)
      process.exit(1)
    }
  }

  /*
  // Clean lib using ts-node
  try {
    const cleanLibArgs = ['npx', 'ts-node', path.join(root, 'util', 'clean-lib.ts'), argv.data]
    if (argv.prefs) cleanLibArgs.push('--prefs')
    if (argv.attachments) cleanLibArgs.push('--attachments')
    execSync(cleanLibArgs.join(' '), { cwd: root, stdio: 'inherit' })
  }
  catch (error) {
    console.error('Error running clean-lib.ts:', error.message)
    process.exit(1)
  }
  */

  // Check for non-existing attachments
  try {
    const data = JSON.parse(fs.readFileSync(argv.data, 'utf8'))
    for (const item of data.items) {
      for (const att of (item.attachments || [])) {
        if (att.path) {
          const attPath = path.join(
            root,
            'test/fixtures/export',
            path.basename(path.dirname(att.path)),
            att.path.replace('STORAGE:', ''),
          )
          if (!fs.existsSync(attPath)) {
            console.warn(`*** WARNING ***: ${argv.data} has non-existing attachment ${att.path}`)
          }
        }
      }
    }
  }
  catch (error) {
    console.error('Error processing data file:', error.message)
    process.exit(1)
  }

  const document = (await read(argv.feature))[0]

  const name = argv.import
    ? /^Import <references> references/
    : new RegExp(`^Export <references> references for ${Translator.name} to`)
  const outlines = document.feature.elements
    .filter(_ => _.keyword === 'Scenario Outline' && _.name.match(name))

  if (outlines.length !== 1) {
    console.error(`Error: Found ${outlines.length} outlines containing "${Translator.name}"`)
    process.exit(1)
  }

  const outline = outlines[0]
  const examples = outlines[0].examples[0].body
  const existing = examples.find(example => example.cells[0].value === argv.title)
  if (argv.export) {
    const data = JSON.parse(fs.readFileSync(argv.data, 'utf8'))
    argv.number = data.items.length
  }
  if (existing) {
    existing.cells[1] = new TableCell(`${argv.number}`)
  }
  else {
    examples.unshift(new TableRow([new TableCell(argv.title), new TableCell(`${argv.number}`)]))
  }
  fs.writeFileSync(argv.feature, format(document))

  // Copy/create test fixtures
  const fixture = path.join(root, `test/fixtures/${argv.mode}`, argv.title)
  const source = `${fixture}.json`
  try {
    fs.copyFileSync(argv.data, source)
    execSync(`git add "${source}"`, { cwd: root, stdio: 'inherit' })

    const target = `${fixture}${argv.export ? Translator.ext : '.bib'}`
    if (Translator.name === 'CSL-JSON') {
      fs.writeFileSync(target, '{}', 'utf8')
    }
    else if (!fs.existsSync(target)) {
      fs.writeFileSync(target, '', 'utf8')
    }

    execSync(`git add "${target}"`, { cwd: root, stdio: 'inherit' })
  }
  catch (err) {
    console.error('Error creating/copying fixture files:', err.message)
    process.exit(1)
  }
}

main()
